doc/api/cli.md

```diff
@@ -871,6 +871,43 @@ added: v11.8.0
 
 Use the specified file as a security policy.
 
+### `--experimental-require-module`
+
+<!-- YAML
+added: REPLACEME
+-->
+
+> Stability: 1.1 - Active Developement
+
+Supports loading a synchronous ES module graph in `require()`.
+
+When `--experimental-require-module` is enabled, and `require()`
+resolves to an ES module, if the module is fully synchronous
+(contains no top-level await), the `require()` call returns the
+module name space object. Otherwise, Node.js throws an error
+without executing the module. If `--experimental-print-required-tla`
+is passed, Node.js will evaluate the module, try to locate the
+top-level awaits, and print their location to help users fix them.
+
+With this option alone, the module being `require()`'d should be
+explicitly marked as an ES module either using the `"type": "module"`
+field in `package.json` or the `.mjs` extension. To load implicit ES
+modules ending with `.js` using automatic syntax-based module type
+detection, use `--experimental-require-module-with-detection`.
+
+### `--experimental-require-module-with-detection`
+
+<!-- YAML
+added: REPLACEME
+-->
+
+> Stability: 1.1 - Active Developement
+
+In addition to what `--experimental-require-module` supports, when the module
+being `require()`'d is not explicitly marked as an ES Module using the
+`"type": "module"` field in `package.json` or the `.mjs` extension, it will
+try to detect module type based on the syntax of the module.
```

Comment:
This needs to be controlled by `--experimental-detect-module`. It’s important that a file be interpreted as the same format by both the CommonJS and ESM loaders or else we might get very unexpected behaviors; we wouldn’t want `import('./file.js')` to think that `file.js` is CommonJS and `require('./file.js')` to think that `file.js` is ESM. Either detection needs to be fully on—affecting both `require` and `import`—or fully off, affecting neither (for this branch, where both `require` and `import` support both CJS and ESM). I think it’s fine to allow toggling detection on/off for `require`, but it should be via the same `--experimental-detect-module` flag that we already have so that users can only enable or disable it for all modules, not just for `require`s or for `import`s individually.

And not only does it need to be on or off for both `require` and `import` together, but both methods need to use the _same_ logic for detection. We can’t have the `require` detection algorithm returning possibly different results than the `import` detection algorithm.

Reply:
When I was testing I actually realized detection isn't that useful for most cases. So I will just put it in a follow-up PR.

Reply:
I _think_ it’s okay to have `require(esm)` not support detection at all while `import` supports it via `--experimental-detect-module`, because `require(esm)` will treat all “ambiguous” files as CommonJS and nothing that parses successfully as CommonJS will be detected as ESM by `--experimental-detect-module`, so there should never be the case where a file is treated as CommonJS by `require` and ESM by `import` or vice versa. But I might be missing a case here.

The safest approach, though, is to just have the one flag apply to both and have the same logic run for both. Why not just do that as part of this PR?

Reply:
I don't want to couple require(esm) with --experimental-detect-module because I think the implementation of the latter has a lot of issues that we talked about before (reading and parsing the file twice etc.) and it seems unfortunate to me that once you enable require(esm) detection you also get yourself into those troubles. Might as well just keep detection out of the scope of this PR. I don't think it needs to cover all these to land as a first iteration experiment. Without detection this is already can load all the ~20 high impact esm-only npm packages that I've tested just fine. Detection only serves a less painful point that can be already be addressed with explicit module type while there's no workaround in core for require(esm) at all ATM. IMO the latter is the major pain point of ESM adoption while the former is minor and less of a priority, so I just want to focus on the higher priority.